### PR TITLE
Add mentionees property to TextMessageEvent

### PR DESCRIPTION
**Pull request by nanato12**
# Overview
Messaging API update for January 2021
https://developers.line.biz/en/news/2021/01/20/messaging-api-update-january-2021/

**Text message type properties**

## Related Issue
- How to mention member with bot? #321
- Messaging API - January 2021 update #330

### CODE REVIEW

File: src/LINEBot/Event/MessageEvent/MentioneeInfo.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+     *
+     * @param array $mentioneeInfo
+     */
+    public function __construct($mentioneeInfo)
+    {
+        $this->index = $mentioneeInfo['index'];
+        $this->length = $mentioneeInfo['length'];
+        if (isset($mentioneeInfo['userId'])) {
+            $this->userId = $mentioneeInfo['userId'];
+        }
```
Comment (nanato12) [inline src/LINEBot/Event/MessageEvent/MentioneeInfo.php:41]: Sometimes can't get a userId.  

スクリーンショット 2021-01-20 14 10 32

Comment (nanato12) [inline src/LINEBot/Event/MessageEvent/MentioneeInfo.php:41]: Sometimes can't get a userId.  
スクリーンショット 2021-01-20 14 10 32

Comment (Vaduz) [reply src/LINEBot/Event/MessageEvent/MentioneeInfo.php:41]: https://developers.line.biz/en/docs/messaging-api/user-consent/
User consent is required to obtain user information.

File: src/LINEBot/Event/MessageEvent/MentioneeInfo.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+namespace LINE\LINEBot\Event\MessageEvent;
+
+class MentioneeInfo
```
Comment (Vaduz) [inline src/LINEBot/Event/MessageEvent/MentioneeInfo.php:21]: Could you change this name to `Mentionee`?
```suggestion
class Mentionee
```

File: src/LINEBot/Event/MessageEvent/MentioneeInfo.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
+<?php
+
+/**
+ * Copyright 2020 LINE Corporation
```
Comment (Vaduz) [inline src/LINEBot/Event/MessageEvent/MentioneeInfo.php:4]:
```suggestion
 * Copyright 2021 LINE Corporation
```

File: src/LINEBot/Event/MessageEvent/TextMessage.php [hunk 4]
```diff
@@ -68,4 +80,14 @@ public function getEmojis()
     {
         return $this->emojis;
     }
+
+    /**
+     * Returns mentioned user info list of the messages.
+     *
+     * @return array
```
Comment (Vaduz) [inline src/LINEBot/Event/MessageEvent/TextMessage.php:87]:
```suggestion
     * @return array|null
```

**Review by Vaduz — CHANGES_REQUESTED** [review]
Please add tests for these changes.

File: src/LINEBot/Event/MessageEvent/MentioneeInfo.php (hunk 2 above)
Comment (nanato12) [reply src/LINEBot/Event/MessageEvent/MentioneeInfo.php:21]: The emoji object is `EmojiInfo`, do you want to change it?

Comment (nanato12) [reply src/LINEBot/Event/MessageEvent/MentioneeInfo.php:21]: If you change it, change here too.
`src/LINEBot/Event/MessageEvent/TextMessage.php` **line 59**

Comment (Vaduz) [reply src/LINEBot/Event/MessageEvent/MentioneeInfo.php:21]: Ah, it might be inconsistent...
Let's leave it then.

**Comment by nanato12** [comment]
> Please add tests for these changes.

Updated test. https://github.com/line/line-bot-sdk-php/pull/329/commits/b639246303b67970cb1e8afc13cfe73c025c25a7

File: tests/LINEBot/EventRequestParserTest.php [hunk 5]
```diff
@@ -62,15 +62,24 @@ class EventRequestParserTest extends TestCase
    "message":{
     "id":"contentid",
     "type":"text",
-    "text":"message (love)",
+    "text":"@example Hello, world! (love)",
```
Comment (Vaduz) [inline tests/LINEBot/EventRequestParserTest.php:65]: Please create a new message for testing mention.

File: tests/LINEBot/EventRequestParserTest.php [hunk 6]
```diff
@@ -587,7 +596,11 @@ public function testParseEventRequest()
             $this->assertEquals('replytoken', $event->getReplyToken());
             $this->assertEquals('contentid', $event->getMessageId());
             $this->assertEquals('text', $event->getMessageType());
-            $this->assertEquals('message (love)', $event->getText());
+            $this->assertEquals('@example Hello, world! (love)', $event->getText());
```
Comment (Vaduz) [inline tests/LINEBot/EventRequestParserTest.php:599]: And please not to reuse existing assertions.

Comment (nanato12) [reply tests/LINEBot/EventRequestParserTest.php]: Sorry, fixed.

File: tests/LINEBot/EventRequestParserTest.php (hunk 5 above)
Comment (nanato12) [reply tests/LINEBot/EventRequestParserTest.php]: I added new `TextMessageEvent` for testing mention.

File: tests/LINEBot/EventRequestParserTest.php [hunk 7]
```diff
@@ -549,6 +549,30 @@ class EventRequestParserTest extends TestCase
...
+   "message": {
+    "id": "325708",
+    "type": "text",
+    "text": "@example Hello, world! (love)",
+    "mention": {
+     "mentionees": [
+      {
+       "index": 0,
+       "length": 8,
+       "userId": "U850014438e..."
```
Comment (Vaduz) [inline tests/LINEBot/EventRequestParserTest.php:571]: Please add mention message without userId data for code coverage.
And this userId should be full length

Comment (nanato12) [reply tests/LINEBot/EventRequestParserTest.php]: OK, updated **userId** of testing mention.
And, added `testing without mentionee's userId` and `testing without mention`.

File: tests/LINEBot/EventRequestParserTest.php [hunk 8]
```diff
@@ -549,6 +549,68 @@ class EventRequestParserTest extends TestCase
...
+   "mode": "active",
+   "timestamp": 1462629479859,
+   "source": {
+    "type": "user",
+    "userId": "U4af4980629..."
+   },
+   "message": {
+    "id": "325708",
+    "type": "text",
+    "text": "@example message without mention"
```
Comment (Vaduz) [inline tests/LINEBot/EventRequestParserTest.php:612]:
```suggestion
    "text": "message without mention"
```

File: tests/LINEBot/EventRequestParserTest.php [hunk 9]
```diff
@@ -549,6 +549,68 @@ class EventRequestParserTest extends TestCase
...
+   }
+  },
+  {
+   "replyToken": "nHuyWiB7yP5Zw52FIkcQobQuGDXCTA",
+   "type": "message",
+   "mode": "active",
+   "timestamp": 1462629479859,
+   "source": {
+    "type": "user",
+    "userId": "U4af4980629..."
```
Comment (Vaduz) [inline tests/LINEBot/EventRequestParserTest.php:584]:
```suggestion
    "userId": "U0123456789abcd0123456789abcdef"
```

File: tests/LINEBot/EventRequestParserTest.php [hunk 10]
```diff
@@ -549,6 +549,68 @@ class EventRequestParserTest extends TestCase
...
```
Comment (Vaduz) [inline tests/LINEBot/EventRequestParserTest.php:607]:
```suggestion
    "userId": "U0123456789abcd0123456789abcdef"
```

**Comment by nanato12** [comment]
Fixed and CI success.

**Review by Vaduz — APPROVED** [review]
Looks good

File: src/LINEBot/Event/MessageEvent/MentioneeInfo.php [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    public function getLength()
+    {
+        return $this->length;
+    }
+
+    /**
+     * Returns User ID of the mentioned user.
+     *
+     * @return string
```
Comment (moririnson) [inline src/LINEBot/Event/MessageEvent/MentioneeInfo.php:67]: https://github.com/line/line-bot-sdk-php/pull/329/files#diff-95c9ef110095446ce4dbe883780d9d49939e957de86bf5a27cbc247dbd9e5687R39

`string|null` ?

File: src/LINEBot/Event/MessageEvent/MentioneeInfo.php [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    public function getIndex()
+    {
+        return $this->index;
+    }
+
+    /**
+     * Returns length of the text of the mentioned user.
+     *
+     * @return string
```
Comment (moririnson) [inline src/LINEBot/Event/MessageEvent/MentioneeInfo.php:57]: https://developers.line.biz/ja/reference/messaging-api/#wh-text

> emojis[].length Number

`int` ?

Comment (nanato12) [reply src/LINEBot/Event/MessageEvent/MentioneeInfo.php]: Sorry, that's right. fixed.

File: src/LINEBot/Event/MessageEvent/MentioneeInfo.php (hunk 11 above)
Comment (nanato12) [reply src/LINEBot/Event/MessageEvent/MentioneeInfo.php]: fixed too

**Review by moririnson — APPROVED** [review]

**Comment by moririnson** [comment]
@nanato12 thanks!

**Pull request merged**